New components

garyb opened the pull request:
:boom:

Resolves #138, resolves #133.

I can squash the commits before merging, and probably should add some example/tests for installed components/child queries etc. but the general API is good to review. Perhaps needs some more doc comments too.

There are a few other changes in here that I probably should have done separately afterwards:

- `Attr` representation has been overhauled to distinguish between `Attr`s and `Prop`s. Also `Key` is a special value now also. The type is called `Prop`, as all the constructors do all represent some kind of property, `Meta` seemed like it might be a bit confusing.
- Not every property has a corresponding attribute, and for those that do the property name and attribute name are not necessarily the same, so the class for rendering properties in the string renderer was adjusted slightly to accommodate this.
- `Element`s can now be namespaced
- Removed all the empty element constructors that previously accepted children (for example, `input_` has been renamed to `input` and there is no equivalent to the old `input` now - `<input>` isn't allowed to have content inside it).
- The form event helpers were renamed to be more consistent with the events (`onValueChanged` -> `onValueChange` corresponding to `onChange`, `onInput` -> `onValueInput` and the "normal" form of `onInput` was added to `Halogen.HTML.Events` as it previously was missing).
- The form event helpers only require an `Applicative f` rather than `Alternative f` now, as we can return `unit` instead of using `empty` (I think this behaviour is ok? as generally it will be `Pure unit` as `f` will be some `Free`).

I removed much of the machinery that we no longer require to make things work, but that include signals, do we want to keep those still? Maybe they should be moved into their own project. Phil also has a version of signal functions implemented on top of `purescript-signal`, so maybe that's enough.

I think that covers it.

jdegoes commented:
:fireworks: :beers: Will review everything in detail tomorrow! Very excited for these changes to land.

puffnfresh commented:
Would love to read a summary of how the changes contribute to component querying and localisation.

jdegoes commented on src/Halogen.purs:
```diff
@@ -1,159 +1,78 @@
+
+type Driver f eff = forall i. f i -> Eff (HalogenEffects eff) i
+
+runUI :: forall eff s f. Component s f (Eff (HalogenEffects eff)) Void
```
This does not allow asynchronous effects. The type should probably be `Aff`, both here and in `Driver`.

jdegoes replied:
Or actually, one could generalize to `MonadAff`, which is the asynchronous equivalent of `MonadEff`.

garyb replied:
Ah right, yes, I wanted to talk about this (`Aff` vs `Eff`, I'll reply to MonadAff separately)

One slightly unfortunate thing about using `Aff` over `Eff` is since Halogen then has to run the `Aff`s it means that if there's a failure and `attempt` wasn't used the error kinda disappears (in the current Halogen we just log it to the console).

If you use `Eff` instead, it means one has to `runAff` themselves, and there's no way that error can disappear, unless `launchAff` was used, which is the expected behaviour anyway.

I'm totally fine with using `Aff`, just wanted to discuss this tradeoff and forgot to mention it in the PR.

garyb replied:
Regarding `MonadAff`/`MonadEff`, I was really hoping that approach would work, as it makes for much nicer types, but the problem is we'd have to provide an `m a -> Eff _ a` function anyway to make it work, as we need a way of running the effect when calling the driver from the event handlers: https://github.com/slamdata/purescript-halogen/blob/c7ecfc88efe71da246bf45b79a69bfe66ae11fa2/src/Halogen/HTML/Renderer/VirtualDOM.purs#L40-L41

garyb replied:
Ah! I think there's nothing to discuss after all, I'm pretty sure it has to be `Aff` rather than `Eff`, as we don't have access to the driver in `eval` for a component.

garyb replied:
I'm working on this now, but it's a bit tricky as we need a `MonadRec` instance for `Aff` (because of `runFreeM`/`runFreeCM`).

jdegoes replied:
`Aff` should have a `MonadRec`, it should probably let the stack grow for a bit before jumping off using `later`.

jdegoes replied:
Let me know if you need me to add `MonadRec` to `Aff` or if you have that (it is something I have been meaning to do, since a valid implementation is possible; a performant one is trickier since you want to stay on the stack for some time before bouncing off using `later`).

garyb replied:
I'm just about to test one now, I let it run 50 times before bouncing off. It's hard to know what a sensible value is!

jdegoes replied:
50 times should be fine (don't want to get too aggressive). I've done call stacks nested hundreds of layers deep in JS.

garyb replied:
Yeah, that's what I figured - 50 seemed conservative yet a reasonable maximum that we'll probably never reach in the Halogen case, or certainly not often enough for it to be a problem.

jdegoes replied:
Consider that Promises/A+ bounce on every bind. :smile:

garyb replied:
Wow, I did not know that! :scream:

We were also talking about using [asap.js](https://github.com/kriskowal/asap) in #purescript, but that would make using aff a bit more complicated with the extra dependency.

natefaubion replied:
Most fast promises library use something similar to asap, which is essentially bouncing like what monad rec does. The promises spec only requires that it happens after the current call stack, not that it happens on a subsequent tick. I've modified some hot algorithms in sweet.js to use Bluebird as an expirement, and only saw about a 10-15% decrease in performance. The raw part of asap.js is small enough to just inline as well. It's certainly not any longer than other FFI code in Aff :)

jdegoes commented on src/Halogen.purs:
```diff
@@ -1,159 +1,78 @@
+  driver ref q = do
+    refVal <- readRef ref
+    case refVal of
+      Nothing -> throwException $ error "Error: An attempt to re-render was made during the initial render."
```
With `Aff`, these would likely change to `AVar`, which would have the nice side-effect of eliminating this bit of unpleasantness.

jdegoes commented on src/Halogen.purs:
```diff
@@ -1,159 +1,78 @@
+            writeRef ref $ Just { node: node', vtree: next, state: s'' }
+            pure i
+
+actionF :: forall f g. (Functor f, Functor g, Inject f g) => (forall i. i -> f i) -> Free g Unit
```
This function and the others could use some documentation. It's not immediately clear what purpose they serve.

jdegoes commented on src/Halogen/Component.purs:
```diff
@@ -1,64 +1,214 @@
+              , query: c.query
+              }
+
+component :: forall s f g p. (s -> HTML p (f Unit)) -> (forall i. f i -> StateT s g i) -> Component s f g p
```
This one is not using the type aliases like the other ones, which makes it inconsistent and its signature overly complex.

garyb replied:
Oops thanks, I only came up with those synonyms today when working on the examples, so forgot about this one.

jdegoes commented on src/Halogen/HTML/Core.purs:
```diff
@@ -0,0 +1,209 @@
+  | Attr (Maybe Namespace) AttrName String
+  | Key String
+  | Handler (ExistsR (HandlerF i))
+  | Initializer (HTMLElement -> i)
```
I like the new initializers and finalizers, should eliminate the need for some post render hooks.

jdegoes commented on src/Halogen/HTML/Events.purs:
```diff
@@ -1,170 +1,129 @@
+import Halogen.HTML.Events.Types (Event(), MouseEvent(), FocusEvent(), KeyboardEvent())
+import Halogen.HTML.Core (Prop(), handler, eventName)
+
+inputF :: forall f g a. (Functor f, Functor g, Inject f g) => (forall i. a -> i -> f i) -> a -> EventHandler (Free g Unit)
```
Are these all useful functions? Would be worthwhile explaining what they're used for.

garyb replied:
Yeah, these along with the `actionF`/`requestF` and `FC` variants all need some explanatory comments. I'm pretty sure they are useful though, as they essentially let you use DSL constructors directly: https://github.com/slamdata/purescript-halogen/blob/5b817583266b03e32622f749dcd6a1173e0d7ed6/test/Test/Main.purs#L43

jdegoes commented on src/Halogen/HTML/Events/Forms.purs:
```diff
@@ -1,48 +1,34 @@
-import Halogen.HTML.Events.Handler
+-- | Attach event handler to event `key` with getting `prop` field as an
+-- | argument of `handler`.
+addForeignPropHandler :: forall f value. (Applicative f, IsForeign value) => String -> String -> (value -> EventHandler (f Unit)) -> Prop (f Unit)
```
This is so terrible, hopefully we can eliminate this someday.

garyb replied:
At least it's private :smile:

paf31 replied:
Curious: what don't you like about it?

jdegoes replied:
Not this function, specifically (since it's private), just the fact that `IsForeign` and friends show up in the public API. :cry:

paf31 replied:
Yeah it's unfortunate. Problem is, you can guess at the fields the event might have, but browser differences etc. mean I wouldn't want to trust what I get back as the event, and want to validate it. This sort of thing probably should be hidden in some low level DOM events library somewhere.

jdegoes replied:
That's what I'm thinking, it would be nice if there were a `purescript-events` library that has an on-opinionated, low-level abstraction across all browsers, which unifies the common stuff and in some type safe fashion supports the useful browser-specific stuff. Halogen and other libs could depend on it.

paf31 replied:
I was working under the assumption that `target.value` gave values of different types based on the input type: checkbox, radio, text, etc., but that's not true. According to the spec, it's always a string: http://www.w3schools.com/jsref/prop_attr_value.asp

So if the implementation is sound and the `value` property isn't set artificially, by a plugin for example, there's no need for a polymorphic `value`. Validating the value is probably still a good idea though, in case of bad plugins or whatever.

jdegoes commented on test/Examples/Intro.purs:
```diff
@@ -0,0 +1,43 @@
+                      [ H.text (if s.on then "On" else "Off") ]
+           ]
+
+  eval :: Eval Input State g
```
Thank god for type aliases. :smile:

jdegoes commented:
> The form event helpers only require an Applicative f rather than Alternative f now, as we can return unit instead of using empty (I think this behaviour is ok? as generally it will be Pure unit as f will be some Free).

Yes, that's fine, in fact, the only type safe way to implement `query` is with `Free`, since we don't have GADTs. You can hide data constructors and (pseudo-safely) take advantage of that knowledge by write `query` without `Free`, but there's going to be coercion involved.

I am glad to see `Event` etc. go, this all seems a lot simpler.

> include signals, do we want to keep those still? Maybe they should be moved into their own project.

Yeah, that should be its own project, it doesn't really depend on Halogen. Signal functions (of HTML) can be compiled into views, so the concept is still useful.

I'll take another look through, but I think most of my (unmade) suggestions can and should be done in separate PRs (for which I'll create separate tickets). For this PR, I'd just like to see a demo of Ace that utilizes nested components. Not only is this one of the most requested examples, but it will prove out the concepts.

Is there any other cleanup-type tasks you passed on for now, or did you get most of your wish list into this PR?

garyb commented:
Great, once the `MonadRec Aff` stuff is in I can finish switching to `Aff` and finish up the ajax example I'm half way through converting, once that is done, I'll do the Ace one.

And nope, I think this is it. Pretty sure I covered all the major things I wanted to clean up in here, aside from maybe the px/% for `width` and `height` values.

jdegoes commented:
@puffnfresh This is basically #138, which explains how the design achieves querying, localization and modularity. However, I think when @garyb does a real Ace example that utilizes local querying and component nesting, it will be much more tangible. If there are any tricky parts you'd like to see in that example, let us know!

garyb commented:
I've switched over to `Aff` from `Eff` now, the `AVars` certainly cleaned up that stuff in `runUI`/`driver`!

What do you think about this? https://github.com/slamdata/purescript-halogen/commit/831681c8624b071d164572b39bba19c3239eeb49#diff-5e398ef5650dd9aaccecdf2f73089d81R49

In the current halogen we just log when there are uncaught errors from an `Aff` that is being run, but I think re-throwing might be more appropriate.

I've also updated the ajax example, and it's revealed an interesting maybe-problem with the new model. When a request is made to the server we actually want to trigger multiple inputs - one to set the UI state to "busy", and then another one that updates it after the `Aff` completes. My first attempt doesn't work, as [I change the busy state during the handler that runs the `Aff`](https://github.com/slamdata/purescript-halogen/commit/831681c8624b071d164572b39bba19c3239eeb49#diff-07dd2ef337cf3a169ba61a5a0f46996eR100), but that doesn't trigger a re-render. I think I know the workaround (combine two inputs), but this wasn't something I entirely expected.

The build is failing right now as I need something to be merged into `purescript-nonempty` and to restructure things a bit so that the examples have their own dependencies.

garyb commented:
Hm, no, "combine two inputs" doesn't help out here either actually, as they still get run "together".

jdegoes commented:
> When a request is made to the server we actually want to trigger multiple inputs - one to set the UI state to "busy", and then another one that updates it after the Aff completes.

Yeah, this is something I have been thinking about (basically the reason we had `Event` in the last model, to emit multiple events). While you can emit multiple commands via `Free`, they will be treated as atomic and interpreted "at the same time" (up to `g`), so it doesn't really suit the above use case. Let me think on this one some more and maybe we can chat later today about solutions.

jdegoes commented:
> In the current halogen we just log when there are uncaught errors from an Aff that is being run, but I think re-throwing might be more appropriate.

That's probably the right thing to do in Javascript, as then the window main uncaught error handler can do whatever it does. Perhaps we should allow users to provide an "uncaught error handler" when they run the UI, we could utilize that to log errors back to SlamData or provide a bug report interface or something like that.

zudov commented:
I see the mixins were removed in https://github.com/slamdata/purescript-halogen/commit/cd7eab837e66aa373653ab1a0cdbd80bf2dbe8cd . What is the plan about those?

garyb commented:
I think something similar should still be possible, perhaps using functor coproducts to do the mixing-in, but I'm not exactly sure yet.

jdegoes commented:
> I think something similar should still be possible, perhaps using functor coproducts to do the mixing-in, but I'm not exactly sure yet.

Yes, you should be able to write combinators that take a component and return another one with a richer query language allowing more commands. 

Although, the old router mixin was not that useful (see purescript-router), and the undo / redo can be done a bit differently, you can actually snapshot & restore the whole app without needing a mixin, since the state is exposed. 

But I'm sure there are plenty of other cases where mixins will be the best solution to some problem.
